src/txorphanage.h

```diff
@@ -43,9 +43,11 @@ class TxOrphanage {
     /** Limit the orphanage to the given maximum */
     unsigned int LimitOrphans(unsigned int max_orphans) EXCLUSIVE_LOCKS_REQUIRED(g_cs_orphans);
 
-    /** Add any orphans that list a particular tx as a parent into a peer's work set
-     * (ie orphans that may have found their final missing parent, and so should be reconsidered for the mempool) */
-    void AddChildrenToWorkSet(const CTransaction& tx, std::set<uint256>& orphan_work_set) const EXCLUSIVE_LOCKS_REQUIRED(g_cs_orphans);
+    /** Which peer provided a parent tx of orphans that need to be reconsidered */
+    std::multimap<NodeId, uint256> m_peer_work_set GUARDED_BY(g_cs_orphans);
```

Comment:
67618f024690bc9b926aa48ead23a0f687f03fe4: I know there is a check to avoid duplicate insertion of the same txid, but is there a reason to pick this datastructure, which doesn't disallow duplicate entries like set? Also insertion is trivially more expensive (log(all peers' work sets) vs log(this peers' work set)). Finally handling is done in insertion order, not in sorted order.

Reply:
> ... is there a reason to pick this datastructure, which doesn't disallow duplicate entries ...

I think the reason could be to allow (nodeX, tx1), (nodeX, tx2)?

Reply:
Note that if you've got N peers each with K items in their work set, then a log search over everything is `log(N*K)`, and a log search for the peer followed by a log search or the work item is `log(N) + log(K)` which is the same. If there's one peer with K items and every other peer has 0, you instead get `log(K)` vs `log(N) + 1/N log(K)`; so a map of sets would be better than a set of pairs provided `N < K`, but probably isn't very interesting.

I picked the data structure because I thought it was simpler than a set of pairs, but I don't think there's an adequate check to avoid duplicate insertion of the same txid (and to do that, you'd have to search through all the entries for a peer, which would be annoying).

Reply:
Changed to `map<NodeId,set<uint256>>`